src/resolve-config.js

```diff
@@ -2,40 +2,48 @@
 
 const cosmiconfig = require("cosmiconfig");
 const minimatch = require("minimatch");
+const mem = require("mem");
 
-const asyncWithCache = cosmiconfig("prettier");
-const asyncNoCache = cosmiconfig("prettier", { cache: false });
-const syncWithCache = cosmiconfig("prettier", { sync: true });
-const syncNoCache = cosmiconfig("prettier", { cache: false, sync: true });
+const getExplorerMemoized = mem(opts =>
+  cosmiconfig("prettier", { sync: opts.sync, cache: opts.cache })
+);
+
+/** @param {{ cache: boolean, sync: boolean }} opts */
+function getLoadFunction(opts) {
+  // Normalize opts before passing to a memoized function
+  opts = Object.assign({ sync: false, cache: false }, opts);
+  return getExplorerMemoized(opts).load;
+}
 
 function resolveConfig(filePath, opts) {
-  const useCache = !(opts && opts.useCache === false);
-  return (useCache ? asyncWithCache : asyncNoCache)
-    .load(filePath)
-    .then(result => {
-      return !result ? null : mergeOverrides(result.config, filePath);
-    });
+  opts = opts || {};
+  const load = getLoadFunction({ cache: !!opts.useCache, sync: false });
+  return load(filePath, opts.config).then(result => {
+    return !result ? null : mergeOverrides(result.config, filePath);
+  });
 }
 
 resolveConfig.sync = (filePath, opts) => {
-  const useCache = !(opts && opts.useCache === false);
-  const result = (useCache ? syncWithCache : syncNoCache).load(filePath);
+  opts = opts || {};
+  const load = getLoadFunction({ cache: !!opts.useCache, sync: true });
+  const result = load(filePath, opts.config);
   return !result ? null : mergeOverrides(result.config, filePath);
 };
 
 function clearCache() {
-  syncWithCache.clearCaches();
-  asyncWithCache.clearCaches();
+  mem.clear(getExplorerMemoized);
```

Comment:
Figured that if we're using `mem` for holding cosmiconfig explorers we may as well just clear the mem cache instead of mutating the explorers.